zerver/lib/idempotent_request.py

```diff
@@ -0,0 +1,210 @@
+
+    3- Duplicate failed work (completed=False), immediately raise
+    the cached error (non-transient error).
+
+    It also handles the case of concurrent requests, by locking
+    the corresponding row during the transaction.
+
+    This decorator expects the decorated function to receive
+    idempotency_key and user as kwargs.
+    """
+
+    def idempotency_decorator(
+        do_work: Callable[ParamT, WorkReturnT],
+    ) -> Callable[ParamT, WorkReturnT]:
+        @wraps(do_work)
+        def wrapper(*args: ParamT.args, **kwargs: ParamT.kwargs) -> WorkReturnT:
+            # Cast it to a dict to access its values like a normal dict and keep mypy happy.
+            # **kwargs (without ParamSpec) is already treated as dict in python anyway.
+            _kwargs = cast(dict[str, Any], kwargs)
+            idempotency_key = _kwargs.get("idempotency_key")
+            user = _kwargs.get("user")
+
+            with transaction.atomic(durable=durable, savepoint=savepoint):
+                # Idempotency-Key is currently optional, so we still do the work normally
+                # in case it's omitted by client.
+                if idempotency_key is None:
+                    return do_work(*args, **kwargs)
+
+                try:
+                    assert user is not None
+                    with connection.cursor() as cursor:
+                        # Here we select the matching row from idempotency table and
+                        # apply a row-lock on it during the whole transaction
+                        # but abort if another request is already holding a lock.
+                        query = SQL("""
+                                    SELECT completed, cached_result FROM {idempotency_table}
+                                    WHERE realm_id = %(realm_id)s
+                                    AND user_id = %(user_id)s
+                                    AND idempotency_key = %(idempotency_key_value)s
+                                    FOR UPDATE NOWAIT;
+                                    """).format(
+                            idempotency_table=Identifier(IdempotentRequest._meta.db_table)
+                        )
+                        cursor.execute(
+                            query,
+                            {
+                                "idempotency_key_value": idempotency_key,
+                                "realm_id": user.realm_id,
+                                "user_id": user.id,
+                            },
+                        )
+                        result = cursor.fetchall()
+                        completed = result[0][0]
+                        cached_result = result[0][1]
+
+                        # Duplicate succeeded work.
+                        if completed:
+                            return cached_result_deserializer(orjson.loads(cached_result))
+
+                        # Duplicate failed work.
+                        if completed is False:  # nocoverage
```

Comment:
This should not have a `nocoverage` -- it's a key part of the implementation.  See below.